sm.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH SM 1 "November 24, 2012"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
sm \- Displays a short text fullscreen
.SH SYNOPSIS
.B sm
[ OPTIONS ]
[
.I text
|
.B \-
]
.SH DESCRIPTION
.BR Screen\ Message 
will display a given multi-line message as large as
possible, fullscreen and black on white. You can specify the text either
when launching sm, or edit it while the program is running.
.PP
After a short timeout, the text entry and the quit button will disappear, leaving
nothing on the screen but the entered text. To continue entering text, just start
typing or (left-)click anywhere on the screen.
.PP
To clear the displayed text, press Escape.
.PP
To invert the colors of the text and the background, press Ctrl-I.
.PP
To quit the program, press Ctrl-Q, or Escape twice, or click the "Quit"-button.
.br

.SH OPTIONS
.TP
[ \fItext\fR | \fB\-\fR ]
Text to display at start up. Defaults to ":-)". If "\fB\-\fR" is passed to sm,
it will read the text to display from the standard input, see \fBREMOTE CONTROLLING SM\fR.
.TP
.BI \-k,\ \-\-kiosk
Non-interactive mode, ignoring user input except for Ctrl-Q.
.TP
.BI \-f,\ \-\-foreground= colordesc
Define a different color to use for the foreground of the text to
display than black. The text string can be in any of the forms accepted
by XParseColor; these include name for a color from rgb.txt, such as
DarkSlateGray, or a hex specification such as #3050b2 or #35b.
.TP
.BI \-b,\ \-\-background= colordesc
Define a different color to use for the background of the text to
display than white. For possible values, see above.
.TP
.BI \-i,\ \-\-invert
Switch the roles for foreground and background colors. Useful if you prefer
white-on-black.
.TP
.BI \-n,\ \-\-font= fontspec
Define a different font to use than the default sans-serif font of your
system. The fontspec be the complete name for a truetype font (like
"DejaVu Sans" or "Bitstream Vera Serif") or just a short font family
specification ("serif", "sans-serif").
.TP
.BI \-r,\ \-\-rotate= rotation
Rotates the display by rotation*90 degrees counter-clock-wise. So \-r 1 
rotates the display to the left, and \-r 2 puts it upside down.
.TP
.BI \-a,\ \-\-align= alignment
Aligns the text centered (\-a 0), left (\-a 1) or right (\-a 2).
.TP
.BI \-m,\ \-\-markup
Interpret input text as Pango markup
XML instead of plain text, with attributes to override styling.
The syntax is described on
<\fIhttps://docs.gtk.org/Pango/pango_markup.html\fR>.
For example:

.in +4
.nf
The <big>quick <span color="#F60">brown</span> &#x1F98A;</big>
jumped over the <span letter_spacing="102400">lazy &#128021;</span>.
.fi
.in

Invalid syntax will result in the unparsed prompt being shown.
.TP
.BI \-\-
(Double dash) End option parsing. This is used to be able to actually hand over text
that starts of with an dash.
.TP
.BI \-h,\ \-\-help
This option will give you a short usage message summarizing the
recognized options and quits.
.TP
.BI \-V,\ \-\-version
This prints the project name together with its version number
quits.

.SH REMOTE CONTROLLING SM
If sm is called with \fB\-\fR as a command line argument, it will read the text to be
shown from the standard input. It will read the input until it reaches the end
of the file, or the form feed character \fB\\f\fR, and show the input read so far at
once. Newline characters at the beginning or the end are ignored. The input is assumed to be UTF-8 encoded 
.RB "(see " utf8 (7)).
.PP
This can be used to create automatic displays of changing data or similar tricks. For example, the following command will create a simple digital watch:
.sp
.in +4
.nf
(while sleep 1; do date +%T; printf \\\\f; done) | sm \-k \-
.fi
.in


.SH AUTHOR
This manual page was written by Joachim Breitner
<mail@joachim-breitner.de> and updated by Gerfried Fuchs <rhonda@deb.at>
to reflect additions for commandline option handling.


CHANGES.md
screen-message (0.29)

  - Add keywords to desktop file as requested as described in #693918

 -- Tiago Bortoletto Vaz <tiago@debian.org>  Wed, 06 Nov 2024 09:33:47 -0500

screen-message (0.27)

  - Move sm from /usr/games to /usr/bin. Thanks to Josh Triplet
  - Move manpage from section 6 to section 1. Thanks to Josh Triplett.

 -- Tiago Bortoletto Vaz <tiago@debian.org>  Wed, 16 Oct 2024 11:24:09 -0400

screen-message (0.26)

  * Fix memory leak (thanks to Santtu Lakkala)

 -- Joachim Breitner <nomeata@debian.org>  Fri, 26 Feb 2021 14:12:32 +0100

screen-message (0.25)

  * Support -i and Ctrl-I

 -- Joachim Breitner <nomeata@debian.org>  Tue, 09 May 2017 08:11:20 -0400

screen-message (0.24)

  * New upstream release

 -- Joachim Breitner <nomeata@debian.org>  Mon, 09 May 2016 12:50:10 +0200

screen-message (0.23) unstable; urgency=medium

  Disable temporary disabling of anti-aliasing

  It seems to have stopped working with recent versions of gtk3, and
  furthermore causes every second keypress to be ignored.

 -- Joachim Breitner <nomeata@debian.org>  Tue, 23 Dec 2014 19:41:09 +0100

screen-message (0.22.2)

  * New upstream release

 -- Joachim Breitner <nomeata@debian.org>  Tue, 30 Sep 2014 15:48:32 +0200

screen-message (0.22.1)

  * New minor upstream release (fixes sm.desktop), thanks to Sebastian Pipping
    for the patch.

 -- Joachim Breitner <nomeata@debian.org>  Fri, 30 May 2014 10:21:26 +0200

screen-message (0.22)

  * New minor upstream release

 -- Joachim Breitner <nomeata@debian.org>  Thu, 22 May 2014 23:54:38 +0200

screen-message (0.21)

  * use stock gtk fullscreen functionality

 -- Joachim Breitner <nomeata@debian.org>  Fri, 09 May 2014 09:45:46 +0200

screen-message (0.20) experimental; urgency=low

  * New upstream release

 -- Joachim Breitner <nomeata@debian.org>  Sat, 24 Nov 2012 20:28:14 +0100

screen-message (0.19) unstable; urgency=low

  * Fixes SIGSEGV by not using g_object_set

 -- Joachim Breitner <nomeata@debian.org>  Mon, 17 May 2010 18:13:34 +0200

screen-message (0.18)

  * Does not focus the hidden text input, avoiding unwanted pop-ups of
    a possible virtual keyboard.
  * Ship .desktop file

 -- Joachim Breitner <nomeata@debian.org>  Fri, 24 Jul 2009 12:46:27 +0200

screen-message (0.17)

  * Properly hide entry when no data is provided

 -- Joachim Breitner <nomeata@debian.org>  Mon, 20 Jul 2009 23:44:46 +0200

screen-message (0.16)

  * 3 second timeout, react on cursor movements, thanks to
    Paul Wise for the patch.
  * Avoid flicker when toggling the entry.

 -- Joachim Breitner <nomeata@debian.org>  Thu, 27 Nov 2008 14:37:00 +0100

screen-message (0.15)

  * Back to using a GtkDraw for the text, behaves better with matchbox
  * Show mouse pointer while entry is visible

 -- Joachim Breitner <nomeata@debian.org>  Sat, 23 Aug 2008 23:50:10 +0200

screen-message (0.14)

  * Only show entry from start on when nothing is provided
  * Redraw screen on screen configuration changes

 -- Joachim Breitner <nomeata@debian.org>  Sat, 16 Aug 2008 16:07:25 -0300

screen-message (0.13)

  * Floats and auto-hides text entry
  * Ignore trailing newlines on stdin, thanks to Chris Lamb for the patch.

 -- Joachim Breitner <nomeata@debian.org>  Tue, 12 Aug 2008 01:35:39 -0300

screen-message (0.12)

  * --foreground, --background, --font, thx to Rhonda
  * --rotate [0|1|2|3] for rotation

 -- Joachim Breitner <nomeata@debian.org>  Tue, 05 Aug 2008 19:39:46 -0300

screen-message (0.11)

  * Manpage in section 6

 -- Joachim Breitner <nomeata@debian.org>  Thu, 17 Apr 2008 23:39:57 +0200

screen-message (0.10) unstable; urgency=low

  * Includes an icon
  * Move binary to /usr/games

 -- Joachim Breitner <nomeata@debian.org>  Thu, 20 Mar 2008 14:15:49 +0100

screen-message (0.9)

  * Starts with text selected

 -- Joachim Breitner <nomeata@debian.org>  Wed, 24 Oct 2007 00:53:43 +0200

screen-message (0.8)

  * Make the initial text size the same as normal text
  * Pressing Esc when there is no text quits the program
    (Both changes thanks to Paul Wise)

 -- Joachim Breitner <nomeata@debian.org>  Mon, 22 Oct 2007 17:53:30 +0200

screen-message (0.7)

  * Fixes: only first argument is used as the text
    Thanks to Florian Ragwitz for the patch.
  * Sets the size to the size of the screen. This makes fullscren
    work with xmonad, please tell me if it causes problems with other
    WMs.

 -- Joachim Breitner <nomeata@debian.org>  Sun, 21 Oct 2007 21:19:41 +0200

screen-message (0.6)

  * Read from stdin if parameter is -
  * Middle-Clicking anywhere works
  * End program upon destroy
  * Fix colors to black/white independent of the theme

 -- Joachim Breitner <nomeata@debian.org>  Tue, 19 Jun 2007 13:19:51 +0100

screen-message (0.5)

  * rename to screen-message, as there ∃ sm in Debian

 -- Joachim Breitner <nomeata@debian.org>  Thu, 14 Jun 2007 20:21:20 +0100

sm (0.4)

  * automakized

 -- Joachim Breitner <nomeata@debian.org>  Thu, 14 Jun 2007 18:07:53 +0100

sm (0.3)

  * Get rid of a warning (thx to Jérémy Bobbio <lunar@debian.org>)
  * Automake the whole thing

 -- Joachim Breitner <nomeata@debian.org>  Thu, 14 Jun 2007 16:51:22 +0100

sm (0.2)

  * Version ready for release

 -- Joachim Breitner <nomeata@debian.org>  Thu, 14 Jun 2007 14:24:19 +0100

sm (0.1)

  * Initial release.

 -- Joachim Breitner <nomeata@debian.org>  Thu, 14 Jun 2007 13:33:54 +0100


sm.c
/*
#     sm.c
#     Copyright (C) 2006 - 2014 Joachim Breitner
#
#     This program is free software; you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation; either version 2 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program; if not, write to the Free Software
#     Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
*/

#include <gtk/gtk.h>
#include <gdk/gdk.h>
#include <pango/pango.h>
#include <cairo.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "config.h"

#ifndef _GNU_SOURCE
#define _GNU_SOURCE                             /* for getopt_long */
#endif
#include <getopt.h>

#define min(x,y) ((x) < (y) ? (x) : (y))

#define AUTOHIDE_TIMEOUT 3

static int timeout_id=0;

static GtkWidget* window;
static GtkWidget* draw;
static GdkCursor *cursor;
static GtkWidget* quit;
static GtkWidget* tv;
static GtkWidget* entry_widget;
static GtkTextBuffer* tb;
static PangoFontDescription *font;
static char *foreground = NULL;
static char *background = NULL;
static int inverted = 0; // 0 = normal, 1 = foreground and background swapped
static GdkRGBA  white, black;
static char *fontdesc = NULL;
static int rotation = 0; // 0 = normal, 1 = left, 2 = inverted, 3 = right
static int alignment = 0; // 0 = centered, 1 = left-aligned, 2 = right-aligned
static int markup = 0; // 0 = plain text, 1 = pango xml
static int interactive = 1;
static GString *partial_input;
static gulong text_change_handler;

gboolean hide_entry(gpointer *user_data) {
	(void)user_data;

	timeout_id = 0;
	gtk_widget_hide(entry_widget);
	gtk_widget_grab_focus(draw);
	gtk_widget_queue_draw(draw);
	gdk_window_set_cursor(gtk_widget_get_window(GTK_WIDGET(draw)), cursor);
	return FALSE;
}

static void show_entry(void) {
	if (timeout_id) {
		g_source_remove(timeout_id);
		timeout_id = 0;
	}
	gtk_widget_show(entry_widget);
	gtk_widget_grab_focus(tv);
	gdk_window_set_cursor(gtk_widget_get_window(GTK_WIDGET(draw)), NULL);

	timeout_id = g_timeout_add_seconds (AUTOHIDE_TIMEOUT, (GSourceFunc)hide_entry, NULL);
}

static void clear_text(GtkAccelGroup *accel, GObject *window, guint keyval,  GdkModifierType modifier) {
	(void)accel;
	(void)window;
	(void)keyval;
	(void)modifier;

	if( gtk_text_buffer_get_char_count(tb) ) {
		gtk_text_buffer_set_text(tb,"",-1);
		show_entry();
	} else {
		gtk_main_quit();
	}
}

static void invert_text(GtkAccelGroup *accel, GObject *window, guint keyval,  GdkModifierType modifier) {
	(void)accel;
	(void)window;
	(void)keyval;
	(void)modifier;

	inverted = !inverted;
	gtk_widget_queue_draw(draw);
}

static char *get_text(void) {
	GtkTextIter start, end;
	gtk_text_buffer_get_start_iter(tb,&start);
	gtk_text_buffer_get_end_iter(tb,&end);
	return gtk_text_buffer_get_text(tb, &start, &end, FALSE);
}


static void redraw(GtkWidget *draw, cairo_t *cr, gpointer data) {
	(void)data;
	gdk_cairo_set_source_rgba(cr, inverted ? &black : &white);
	cairo_paint(cr);

	char *text = get_text();
	if (text && *text) {
		int w1, h1;
		static PangoLayout* layout;
		PangoAttrList *attr = 0;

		if (markup == 1) {
			char *untagged;
			GError *error = NULL;

			if (pango_parse_markup(text, -1, 0, &attr, &untagged, 0, &error)) {
				g_free(text);
				text = untagged;
			} else {
				fprintf(stderr, "Markup ignored: %s\n", error->message);
				g_error_free(error);
			}
		}

		layout = gtk_widget_create_pango_layout(draw, text);
		pango_layout_set_font_description(layout, font);

		switch(alignment){
			case 0: // center
				pango_layout_set_alignment(layout,PANGO_ALIGN_CENTER);
				break;
			case 1: // left
				pango_layout_set_alignment(layout,PANGO_ALIGN_LEFT);
				break;
			case 2: // left
				pango_layout_set_alignment(layout,PANGO_ALIGN_RIGHT);
				break;
			default:
				// we probably don't want to annoy the user, so default to
				// the old default-behaviour:
				pango_layout_set_alignment(layout,PANGO_ALIGN_CENTER);
		}

		if (attr) {
			pango_layout_set_attributes(layout, attr);
			pango_attr_list_unref(attr);
		}

		pango_layout_get_pixel_size(layout, &w1, &h1);
		if (w1>0 && h1>0) {
			int w2 = gtk_widget_get_allocated_width(draw);
			int h2 = gtk_widget_get_allocated_height(draw);

			int rw1, rh1;
			if (rotation == 0 || rotation == 2) {
				rw1 = w1;
				rh1 = h1;
			} else {
				rw1 = h1;
				rh1 = w1;
			}

			double s = min ((double)w2/rw1, (double)h2/rh1);

			if (alignment == 1) { // left align
				cairo_translate(cr, (s * rw1)/2, h2/2);
			} else if (alignment == 2) { // right align
				cairo_translate(cr, w2 - (s * rw1)/2, h2/2);
			} else {
				cairo_translate(cr, w2/2, h2/2);
			}
			cairo_rotate(cr, rotation * G_PI_2);
			cairo_scale(cr, s, s);
			cairo_translate(cr, -w1/2, -h1/2);
			gdk_cairo_set_source_rgba(cr, inverted ? &white : &black);
			pango_cairo_show_layout (cr, layout);
		}
		g_object_unref(layout);
	}
	g_free(text);
}

static gboolean text_keypress(GtkWidget *widget, GdkEventButton *event, gpointer *user_data) {
	(void)widget;
	(void)user_data;

	// forward signal to the text view
	gboolean ret;
	g_signal_emit_by_name(tv, "key-press-event", event, &ret);
	gtk_widget_grab_focus(tv);
	return ret;
}

static gboolean text_clicked(GtkWidget *widget, GdkEventButton *event, gpointer *user_data) {
	(void)widget;
	(void)user_data;

	show_entry();
	if (event->type == GDK_BUTTON_PRESS && event->button == 2) {
		GtkClipboard *cb = gtk_clipboard_get(GDK_SELECTION_PRIMARY);

		gchar *txt = gtk_clipboard_wait_for_text(cb);
		if (txt != NULL) {
			gtk_text_buffer_set_text(tb,txt,-1);
			g_free(txt);
		}

	}
	return FALSE;
}

static gboolean read_chan(GIOChannel *chan, GIOCondition condition, gpointer data){
	(void)condition;
	(void)data;

	gchar buf[1024];
	GString *input;
	GIOStatus stat = G_IO_STATUS_NORMAL;
	gsize read;
	GError *err = NULL;

	while ((stat = g_io_channel_read_chars(chan, buf, sizeof(buf), &read, &err)) == G_IO_STATUS_NORMAL && err == NULL) {
		g_string_append_len(partial_input, buf, read);
	}

	if (err != NULL)
	{
	    fprintf (stderr, "Unable to read stdin: %s\n", err->message);
	    g_error_free (err);
	    return TRUE;
	}


	if (stat == G_IO_STATUS_EOF) {
		// There is an end of file, so use the whole input
		input = g_string_new_len(partial_input->str, partial_input->len);
		g_string_truncate(partial_input, 0);
	} else {
		// There is no end of file. Check for form feed characters.
		// Use from the second-to-last to the last.
		char *last_ff = strrchr(partial_input->str, '\f');
		if (last_ff) {
			*last_ff = '\0';
			char *snd_last_ff = strrchr(partial_input->str, '\f');
			if (snd_last_ff == NULL) snd_last_ff = partial_input->str;
			input = g_string_new_len(snd_last_ff, last_ff - snd_last_ff);
			g_string_erase(partial_input, 0, last_ff - partial_input->str + 1);
		} else {
			return TRUE;
		}
	}

	// remove beginning and trailing newlines, if any
	gsize cnt = 0;
	while ((input->len > cnt) && (input->str[cnt] == '\n')) {
		cnt++;
	}
	g_string_erase(input, 0, cnt);

	while ((input->len > 0) && (input->str[input->len - 1] == '\n')) {
		g_string_truncate(input, input->len - 1);
	}

	if (text_change_handler)
		g_signal_handler_block (tb, text_change_handler);
	gtk_text_buffer_set_text (tb, input->str, input->len);
	if (text_change_handler)
		g_signal_handler_unblock (tb, text_change_handler);

	g_string_free(input, TRUE);

	if (stat == G_IO_STATUS_AGAIN)
		return TRUE;
	else
		return FALSE;
}

static void newtext(void) {
	gtk_widget_queue_draw(draw);
}

static void newtext_show_input(void) {
	show_entry();
}

static void move_cursor(GtkTextView* tv, GtkMovementStep step, gint count, gboolean extend_selection, gpointer user_data) {
	(void)tv;
	(void)step;
	(void)count;
	(void)extend_selection;
	(void)user_data;

	show_entry();
}

static struct option const long_options[] =
{
	{"help",       no_argument,       NULL, 'h'},
	{"version",    no_argument,       NULL, 'V'},
	{"foreground", required_argument, NULL, 'f'},
	{"background", required_argument, NULL, 'b'},
	{"invert",     no_argument,       NULL, 'i'},
	{"font",       required_argument, NULL, 'n'},
	{"rotate",     required_argument, NULL, 'r'},
	{"align",      required_argument, NULL, 'a'},
	{"markup",     no_argument,       NULL, 'm'},
	{"kiosk",      no_argument,       NULL, 'k'},
	{0,0,0,0}
};

static void usage(char *cmd) {
	printf("Usage: %s [-h|--help] [-V|--version] [-f|--foreground=colordesc] [-b|--background=colordesc] [-i|--invert] [-n|--font=fontdesc] [-r|--rotate=0,1,2,3] [-a|--align=0,1,2] [-m|--markup] [-k|--kiosk]\n", cmd);
}

static void version(void) {
	printf("%s\n", PACKAGE_STRING);
}

static void my_gdk_rgba_parse (GdkRGBA *rgba, const gchar *spec) {
	gboolean ret = gdk_rgba_parse(rgba, spec);
	if (!ret)  {
	    fprintf (stderr, "Failed to parse color specification %s\n", spec);
	}
}

int main(int argc, char **argv) {
	GString *input;
	int c;
	int input_provided = 0;

	while ((c = getopt_long (argc, argv, "hVf:b:n:r:a:ikm", long_options, (int *) 0)) != EOF) {
		switch (c) {
			case 'h':
				usage(argv[0]);
				return 0;
				break;

			case 'V':
				version();
				return 0;
				break;

			case 'f':
				foreground = optarg;
				break;
			case 'b':
				background = optarg;
				break;

			case 'n':
				fontdesc = optarg;
				break;
			case 'r':
				rotation = atoi(optarg);
				break;
			case 'a':
				alignment = atoi(optarg);
				break;
			case 'i':
				inverted = !inverted;
				break;
			case 'm':
				markup = !markup;
				break;

			case 'k':
				interactive = !interactive;
				break;

			default:
				/* unknown switch received - at least
				 * give usage but continue and use the
				 * data */
				usage(argv[0]);
				break;
		}
	}

	gtk_init(&argc, &argv);

	window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
	gtk_window_set_icon_name (GTK_WINDOW (window), "sm");
	gtk_window_fullscreen(GTK_WINDOW(window));

	g_signal_connect(window, "destroy", G_CALLBACK(gtk_main_quit), NULL);

	if (foreground != NULL) {
		my_gdk_rgba_parse(&black, foreground);
	} else {
		my_gdk_rgba_parse(&black, "black");
	}
	if (background != NULL) {
		my_gdk_rgba_parse(&white, background);
	} else {
		my_gdk_rgba_parse(&white, "white");
	}

	draw = gtk_drawing_area_new();
	gtk_widget_set_events(draw, GDK_BUTTON_PRESS_MASK|GDK_KEY_PRESS_MASK);
	gtk_widget_set_size_request(draw,400,400);
	if (interactive) {
		g_signal_connect(G_OBJECT(draw), "button-press-event", G_CALLBACK(text_clicked), NULL);
		g_signal_connect(G_OBJECT(draw), "key-press-event", G_CALLBACK(text_keypress), NULL);
		gtk_widget_set_can_focus(draw, TRUE);
	}

	cursor = gdk_cursor_new_for_display(gtk_widget_get_display(GTK_WIDGET(draw)), GDK_BLANK_CURSOR);

	tv = gtk_text_view_new();
	tb = gtk_text_view_get_buffer(GTK_TEXT_VIEW(tv));

	partial_input = g_string_new("");

	if (argc > optind)
		if (!strcmp(argv[optind], "-") ) {
			// read from stdin
			GIOChannel *chan = g_io_channel_unix_new(0);
			g_io_channel_set_flags(chan, G_IO_FLAG_NONBLOCK, NULL);
			g_io_add_watch (chan, G_IO_IN | G_IO_HUP, &read_chan, NULL);

			input = g_string_new("");
			input_provided++;
		} else {
			int i;

			input = g_string_new("");

			for (i = optind; i < argc; i++) {
				g_string_append(input, argv[i]);

				if (i < argc - 1) {
					g_string_append(input, " ");
				}
			}
			input_provided++;
		}
	else
		input = g_string_new(":-)");

	gtk_text_buffer_set_text(tb, input->str, input->len);
	g_string_free(input, TRUE);
	GtkTextIter start, end;
	gtk_text_buffer_get_bounds(tb, &start, &end);
	gtk_text_buffer_select_range(tb, &start, &end);

	quit = gtk_button_new_from_icon_name("application-exit", GTK_ICON_SIZE_BUTTON);
	g_signal_connect(G_OBJECT(quit), "clicked", G_CALLBACK(gtk_main_quit), NULL);

	GtkWidget *vbox_button = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
	gtk_box_pack_end(GTK_BOX(vbox_button), quit, FALSE, FALSE, 0);

	GtkWidget *hbox = gtk_box_new(GTK_ORIENTATION_HORIZONTAL,0);
	entry_widget = hbox;
	gtk_box_pack_start(GTK_BOX(hbox), tv,   TRUE,  TRUE,  0);
	gtk_box_pack_start(GTK_BOX(hbox), vbox_button, FALSE, FALSE, 0);

	GtkWidget *vbox = gtk_box_new(GTK_ORIENTATION_VERTICAL,0);
	gtk_box_pack_start(GTK_BOX(vbox), draw, TRUE, TRUE, 0);
	gtk_box_pack_end(GTK_BOX(vbox), hbox, FALSE, FALSE, 0);

	gtk_container_add(GTK_CONTAINER(window), vbox);

	font = pango_font_description_new();
	if (fontdesc != NULL) {
		pango_font_description_set_family(font, fontdesc);
	} else {
		pango_font_description_set_family(font, "sans-serif");
	}
	pango_font_description_set_size(font, 200*PANGO_SCALE);

	GtkAccelGroup *accel = gtk_accel_group_new();
	guint key;
	GdkModifierType mod;
	gtk_accelerator_parse("<Ctrl>Q", &key, &mod);
	gtk_accel_group_connect(accel, key, mod, 0, g_cclosure_new(G_CALLBACK(gtk_main_quit), NULL, NULL));
	if (interactive) {
		gtk_accelerator_parse("Escape", &key, &mod);
		gtk_accel_group_connect(accel, key, mod, 0, g_cclosure_new(G_CALLBACK(clear_text), NULL, NULL));
		gtk_accelerator_parse("<Ctrl>I", &key, &mod);
		gtk_accel_group_connect(accel, key, mod, 0, g_cclosure_new(G_CALLBACK(invert_text), NULL, NULL));
	}
	gtk_window_add_accel_group(GTK_WINDOW(window), accel);
	gtk_widget_show_all(window);

	g_signal_connect_after(G_OBJECT(draw), "draw", G_CALLBACK(redraw), NULL);
	g_signal_connect(G_OBJECT(tb), "changed", G_CALLBACK(newtext), NULL);
	if (interactive) {
		text_change_handler = g_signal_connect(G_OBJECT(tb), "changed", G_CALLBACK(newtext_show_input), NULL);
		g_signal_connect(G_OBJECT(tv), "move-cursor", G_CALLBACK(move_cursor), NULL);
	}

	if (input_provided || !interactive)
		hide_entry(NULL);
	else
		show_entry();

	gtk_main();

	return 0;
}
